### PR TITLE
Grid summary/pinned data fixes (Fixes #1289, #1290)

### LINKED ISSUE #1289: Store summaryData cannot be updated without triggering reactions on records
Store.updateData will always update its RecordSets even if the updated data is empty. This is especially problematic for apps which want to update the summary data when the records change (due to filtering), which results in mobx errors because of unstable reactions.

### PR DESCRIPTION
**Pull request by haynesjm42**
Hoist P/R Checklist
-------------------
Review and check off the below. Items that do not apply should be checked to indicate they have been considered. If unclear if a step is relevant, please leave unchecked and note in comments.

- [x] Up to date with `develop` branch as of last change.
- [x] Ready for review (or add `wip` label).
- [x] Added CHANGELOG entry (or N/A)
- [x] Checked for breaking changes (add `breaking-change` label + CHANGELOG if so, or N/A)
- [N/A] Updated doc comments / prop-types (or N/A)
- [N/A] Reviewed and tested on Mobile (or N/A)
- [N/A] Created Toolbox branch / PR (link provided here, or N/A)

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to collapse multiple intermediate commits into a single commit representing the overall feature change. This helps keep the commit log clean and easy to scan across releases.

Restores support for application usage of the ag-Grid pinned data apis and fixes potential issuers where an application might want to update summary data from a reaction on the Store `records` property.

We now flag the summary record with `xhIsSummary=true` so the Grid can identify it when updating the pinned row data if the summary row is enabled. Grid will extract it from the existing pinned row data and place it back in the appropriate place (top-most for top pinned, bottom-most for bottom pinned). I also moved updating of the pinned data to its own reaction instead of using the data reaction as it was unnecessary to do all the extra work that is done in data reaction and ti seemed like a valid/common use-case where applications would want to update the summary data whenever the Store filter is applied to the `records`, which could/would result in an extra data reaction after the summary data was updated.

In `Store.updateData()` we also now avoid updating our RecordSet references and `lastUpdated` property unless the supplied `rawData` is not empty. This will allow applications to update the summary data with a call like `store.updateData([], newSummaryData)` without causing an infinite loop of reactions.

### CODE REVIEW

**Comment by amcclain** [comment]
Thanks, John - I will be sure to review this later this afternoon and we can get it in soon so @johnbc can move forwards with his use-case.

File: cmp/grid/Grid.js [hunk 1]
```diff
@@ -518,20 +531,24 @@ export class Grid extends Component {
...
+            pinnedTopRowData = agGridModel.getPinnedTopRowData().filter(filterSummaryFn),
+            pinnedBottomRowData = agGridModel.getPinnedBottomRowData().filter(filterSummaryFn);
 
         if (showSummary && store.summaryRecord) {
-            const arr = (showSummary === 'bottom') ? pinnedBottomRecords : pinnedTopRecords;
-            arr.push(store.summaryRecord);
+            if (showSummary === 'bottom') {
+                pinnedBottomRowData.push(store.summaryRecord);
+            } else {
+                pinnedTopRowData.splice(0, 0, store.summaryRecord);
```
Comment (amcclain) [inline cmp/grid/Grid.js]: Just use `unshift()` here?

**Comment by lbwexler** [comment]
So john and I discussed many of the same items (including unshift  :-)) earlier on friday.  I usually mention when we do a verbal code review to avoid any double work -- will be more careful.  In this case, the more eyes the better and we certainly missed the issue about the isSummary() getter.   I am all in favor of not duplicating that state.

**I had originally told John to go ahead and remove all the extra code around skipping the change when we pass in updates = [] to updateData**.  But I now think it is important, perhaps for different reason than we had for starting this bit of the project.  Namely, have been seeing in a client app that running our full data reaction -- even when there are no new records -- is actually pretty expensive. We of course need to understand that, but  I think its a common use case, to do empty updates with a streaming api, so we will want to do this simple optimization.

Agree updateData should always change if this method does anything.  John and I also discussed that point

**Comment by haynesjm42** [comment]
Thanks for the careful look guys, implemented the requested changes. Still don't know how I was completely ignorant of the existence of `unshift` ...

**Comment by lbwexler** [comment]
Thanks John.

I agree with Anselm that we should now change that getter on record to look at the new flag we are setting!  Its both self-documenting, and enhances that getter for cases like this.  Once a summary, always a summary!

**Comment by lbwexler** [comment]
ah -- I see John already whacked it.  OK.  I'll just merge -- we can bring it back if we ever need it!

**Comment by lbwexler** [comment]
don't be so hard on yourself john -- I think I ignored it for a few years
myself -- with a name like that, it doesn't sound like the kind of method
you really need for writing applications.....  :-)

**Comment by amcclain** [comment]
Thank you - this looks like it landed in a very good place.

Yes - the key bit about avoiding the full data reaction is key - even though it's resolved to what we would think is a no-op transaction, it is not free.

**Pull request merged**